addons/pkg/util/condition_util.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package util
+
+import (
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	clusterapiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+
+	"github.com/vmware-tanzu/carvel-kapp-controller/pkg/apis/kappctrl/v1alpha1"
+)
+
+// SummarizeAppConditions summarizes the provided conditions slice into a single condition with the following logic:
+// - If there be any 'Reconciling' condition type with status 'True', the summary condition type will be 'Reconciling'
+// - Otherwise, if there be any 'ReconcileFailed' condition type with status 'True', the summary condition type will be 'ReconcileFailed'
+// - Otherwise, if there be any 'ReconcileSucceeded' condition type with status 'True', the summary condition type will be 'ReconcileSucceeded'
+// - Otherwise, if there be any 'Deleting' condition type with status 'True', the summary condition type will be 'Deleting'
+// - Otherwise, if there be any 'DeleteFailed' condition type with status 'True', the summary condition type will be 'DeleteFailed'
+// - Otherwise, the summary condition type will be 'UnknownCondition'
+// Note that ReconcileFailed|Reconciling|ReconcileSucceeded|Deleting|DeleteFailed|DeleteSucceeded are mutually exclusive
+func SummarizeAppConditions(conditions []v1alpha1.AppCondition) v1alpha1.AppCondition {
+	summaryCondition := v1alpha1.AppCondition{}
+
+	wantedCondTypes := []v1alpha1.AppConditionType{
+		v1alpha1.Reconciling,
+		v1alpha1.ReconcileFailed,
+		v1alpha1.ReconcileSucceeded,
+		v1alpha1.Deleting,
+		v1alpha1.DeleteFailed,
+	}
+	for _, wantedCondType := range wantedCondTypes {
+		for _, cond := range conditions {
+			if cond.Type == wantedCondType && cond.Status == corev1.ConditionTrue {
+				return cond
+			}
+		}
+	}
+	return summaryCondition
```

Comment:
Q: I don't know a lot about kapp conditions but is it correct that if all conditions are false the correct summary is an empty AppCondition?

Reply:
True, which I then check in ```pkgiCondition.Type == "" ``` inside packageinstallstatus_controller.go code for ignoring that condition.
Cause if the condition for ```pkgi``` is False, I didn't see any reason to reflect it inside ```clusterbootstrap``` conditions. Instead, would wait until the ```Status``` filed of the condition changes to True. Does that look flawed to you in any way @sbueringer ?

Reply:
Logic makes sense. I would probably make it more obvious in the code.

Some options:
* return `*v1alpha1.AppCondition` and then in that case `nil` (usually a bit nicer to check against nil vs. having to check that a field is empty)
* rename `summaryCondition` to `emptyCondition` in `SummarizeAppConditions`

Reply:
Agreed, will change. Thanks for your suggestion!